### ISSUE TITLE
Multi-OS with partial docker builds does not run apt stage

### ISSUE THREAD

**Issue by bilderbuchi** [travis-ci/travis-ci#4592]
I have a build in a multi-OS repo where some builds are on the legacy infra, and some on the docker infrastructure. 

*Expected behaviour*: In the docker builds, I can install packages via addons.apt.packages
*Observed behaviour*: In the docker builds, I *cannot* install packages via addons.apt.packages. See #4557 for some previous discussion.

I have tried to reproduce this in a testing repo with the same .travis.yml, but there this works as expected! The only difference I can discern w.r.t. Travis is that the testing repo **has not been enabled for multi-OS**, so I suspect that the problem could lie there.

Original repo build that *does not run apt* [here](https://travis-ci.org/openframeworks/openFrameworks/jobs/73056527#L95), accompanying .travis.yml [here](https://github.com/bilderbuchi/openFrameworks/blob/2b16636f642c8e10012e95568ceb03751fff0edf/.travis.yml).
Test repo build which works *correctly* [here](https://travis-ci.org/bilderbuchi/travis_production_test/jobs/73205154#L76) (never mind that the build fails, this is all later in the build process), with identical .travis.yml [here](https://github.com/bilderbuchi/travis_production_test/blob/51a41fa724c5ec944586753243ec692d62cf917f/.travis.yml).

Second, minor, bug: [Apparently](https://github.com/travis-ci/travis-ci/issues/4557#issuecomment-125670406) the order of `sudo` and `env` in `matrix.include` is significant for the parser to correctly pick up an apt stage. Tweaking this does not resolve the main bug, though.

ping @BanzaiMan who previously helped analyze this.

**Comment by rkh** [comment]
@BanzaiMan any insight? :)

**Comment by BanzaiMan** [comment]
I suspect we are dropping the `addons` key when we expand the multi-os build configuration into a matrix. I haven't looked at this part of the code in a while (not exactly a scenic route), so I don't have any immediate insight.

**Comment by bilderbuchi** [comment]
Is that part of the code public, and if so, where?

**Comment by BanzaiMan** [comment]
If my memory serves me correctly, it is in https://github.com/travis-ci/travis-core/blob/cf7c911312e935873af87b0f8fd5c157c80ffb25/lib/travis/model/build/config/matrix.rb#L14.

Corresponding specs are here: https://github.com/travis-ci/travis-core/blob/cf7c911312e935873af87b0f8fd5c157c80ffb25/spec/travis/model/build/matrix_spec.rb.

**Comment by bilderbuchi** [comment]
Another bit of information to maybe narrow doen the parsing failure: 
If I remove the root-level `os: linux` entry, like [this](https://github.com/bilderbuchi/openFrameworks/commit/ec3750efe01a4d5fce24189d91310af84f1095e7), the APT stage [becomes active](https://travis-ci.org/openframeworks/openFrameworks/jobs/74217980#L95).

**Comment by BanzaiMan** [comment]
I just realized that this is not so much a bug with matrix expansion, but that the `apt` addon is not allowed in PRs.

**Comment by BanzaiMan** [comment]
@bilderbuchi Could you restart https://travis-ci.org/openframeworks/openFrameworks/jobs/73056527 if possible and see if the issue persists? If it does, could you open a new PR and see if the APT addon runs on the PR build? Thanks.

**Comment by bilderbuchi** [comment]
@BanzaiMan I can't restart, git doesn't find the `merge` ref. I'm guessing that is because the PR has been merged in the meantime:
```
$ git fetch origin +refs/pull/4167/merge:

fatal: Couldn't find remote ref refs/pull/4167/merge
```

I'll try to do it in a test repo.

**Comment by bilderbuchi** [comment]
> I'll try to do it in a test repo.

Of course, I never could repro this in a test repo, see first post, so I can't check if the issue still persists currently. :disappointed:
